hypothesis-python/tests/conjecture/common.py

```diff
@@ -67,3 +71,153 @@ def accept(f):
             )
 
     return accept
+
+
+def fresh_data():
+    return ConjectureData(BUFFER_SIZE, prefix=b"", random=Random())
+
+
+@st.composite
+def draw_integer_kwargs(
+    draw,
+    *,
+    use_min_value=True,
+    use_max_value=True,
+    use_shrink_towards=True,
+    use_weights=True,
+    use_forced=False,
+):
+    min_value = None
+    max_value = None
+    shrink_towards = 0
+    weights = None
+
+    # this generation is complicated to deal with maintaining any combination of
+    # the following invariants, depending on which parameters are passed:
+    #
+    # (1) min_value <= forced <= max_value
+    # (2) max_value - min_value + 1 == len(weights)
+    # (3) len(weights) <= 1024
+
+    forced = draw(st.integers()) if use_forced else None
+    if use_weights:
+        # handle the weights case entirely independently from the non-weights
+        # case. We'll treat the weight width here as our "master" draw and base
+        # all other draws around that result.
+        assert use_max_value
+        assert use_min_value
+
+        width = draw(st.integers(1, 1024))
+        weights = draw(
+            st.lists(
+                # weights doesn't play well with super small floats.
+                st.floats(
+                    min_value=0.1, max_value=1, allow_nan=False, allow_infinity=False
+                ),
+                min_size=width,
+                max_size=width,
+            )
+        )
+        center = forced if use_forced else draw(st.integers())
+        # pick a random pivot point in the width to split into left and right
+        # segments, for min and max value.
+        pivot = draw(st.integers(0, width - 1))
+        min_value = center - pivot
+        max_value = center + (width - pivot - 1)
```

Comment:
I prefer to treat the _weights list_ rather than the number of weights as the key attribute, and we can simplify further by using the offset from `min_value` and observing that finite bounds implies disallowing nonfinite floats.

```suggestion
        # handle the weights case entirely independently from the non-weights case.
        # We'll treat the weights as our "master" draw and base all other draws on that.
        assert use_max_value
        assert use_min_value
        
        # weights doesn't play well with super small floats, so exclude <.1
        weights = draw(st.lists(st.floats(0.1, 1), min_size=1, max_size=1024))
        
        # we additionally pick a central value (if not forced), and then the index into
        # the weights at which it can be found - aka the min-value offset.
        center = forced if use_forced else draw(st.integers())
        min_value = center - draw(st.integers(-(len(weights) - 1), len(weights) - 1
        max_value = min_value + len(weights)
```

Reply:
committed a variant of this 👍